google/cloud/internal/rest_carrier.cc

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/internal/rest_carrier.h"
+#include "google/cloud/internal/rest_context.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+opentelemetry::nostd::string_view RestCarrier::Get(
+    opentelemetry::nostd::string_view) const noexcept {
+  // Since the client is never extracting data from the REST headers, we are not
+  // implementing Get and returning an empty string by default.
+  return "";
+}
+
+void RestCarrier::Set(opentelemetry::nostd::string_view key,
+                      opentelemetry::nostd::string_view value) noexcept {
+  context_.AddHeader(std::string(key.data()), std::string(value.data()));
```

Comment:
"using `string_view::data()` without `string_view::size()` is a red flag" - @devbww.

`string_view`s do not need to be `\0` terminated. So checking `data()` is not enough.

```cc
  internal::NoExceptAction([=] {
    context_.AddHeader(std::string(key.data(), key.value()),
                       std::string(value.data(), value.size()));
  });
```

Reply:
Changed. Can you clarify the string_view thing? "app" as a std::string is ["a","p","p", "\0"] but as a string_view ["a", "p", "p"]?

Reply:
It is probably best to forget about any NUL beyond the end of a `std::string`.  That is, `char const* p = "app"` is `['a', 'p', 'p', '\0']`, but `std::string s(p)` and `absl::string_view(p)` are just `['a', 'p', 'p']`.

But then consider `absl::string_view(p, 1)`.  How could you convert that to a `std::string`?

Reply:
I do not know how to link it, but look at the notes for constructor (5) here: https://en.cppreference.com/w/cpp/string/basic_string/basic_string

"The length of the string is determined by the first null character."

If the string_view is not null-terminated, we are going to have a bad time.

---

If I say:
```cc
char* data = "app";
```

That is equivalent to:
```cc
char data[] = {'a', 'p', 'p', '\0'};
```

And converting that to a `std::string` would be fine.

But here is a test that exercises the non-null terminated `string_view` case in some other piece of code:

https://github.com/googleapis/google-cloud-cpp/blob/a9174ce92bb7563cd2f587b2037dc12dc1e1d4db/google/cloud/internal/grpc_opentelemetry_test.cc#L78-L89

Reply:
Based on what @dbolduc wrote and to answer @devbww:

```
absl::string_view sv = absl::string_view(p, 1)  // is "{'a'}
```

stores a pointer to the character array and the length of the substring, so to convert it to a string:

```
std::string s = std::string(sv.data()) // bad, because the string doesn't have a null-terminated character
std::string s = std::string(sv.data(), sv.size())` // creates "a" at the correct length
```

So we use the size to create a string, since the string view won't always be null terminated.